Studio/src/Application/Data/MeshWarper.cpp

```diff
@@ -71,17 +103,120 @@ void MeshWarper::check_warp_ready()
   }
 
   // perform warp
-  this->vertices_ = MeshUtils::distilVertexInfo(this->reference_mesh_);
-  this->faces_ = MeshUtils::distilFaceInfo(this->reference_mesh_);
   this->points_ = Eigen::Map<const Eigen::VectorXd>(
     (double*) this->reference_particles_.data_block(),
     this->reference_particles_.size());
   this->points_.resize(3, this->reference_particles_.size() / 3);
+  this->points_.transposeInPlace();
+
+  this->find_good_particles();
+  this->points_ = this->remove_bad_particles(this->points_);
+
+  this->add_particle_vertices();
+
+  this->vertices_ = MeshUtils::distilVertexInfo(this->reference_mesh_);
+  this->faces_ = MeshUtils::distilFaceInfo(this->reference_mesh_);
   this->warp_ = MeshUtils::generateWarpMatrix(this->vertices_, this->faces_,
-                                              this->points_.transpose());
+                                              this->points_);
   this->needs_warp_ = false;
 }
 
+//---------------------------------------------------------------------------
+void MeshWarper::add_particle_vertices()
+{
+
+  for (int i = 0; i < this->points_.rows(); i++) {
+    this->reference_mesh_->BuildLinks();
+    this->reference_mesh_->BuildCells();
+
+    auto locator = vtkSmartPointer<vtkCellLocator>::New();
+    locator->SetCacheCellBounds(true);
+    locator->SetDataSet(this->reference_mesh_);
+    locator->BuildLocator();
+
+    double pt[3]{this->points_(i, 0), this->points_(i, 1), this->points_(i, 2)};
+    double closest_point[3];//the coordinates of the closest point will be returned here
+    double closest_point_dist2; //the squared distance to the closest point will be returned here
+    vtkIdType cell_id; //the cell id of the cell containing the closest point will be returned here
+    int sub_id; //this is rarely used (in triangle strips only, I believe)
+    locator->FindClosestPoint(pt, closest_point, cell_id, sub_id, closest_point_dist2);
+
+    // grab the closest cell
+    vtkCell* cell = this->reference_mesh_->GetCell(cell_id);
+    vtkSmartPointer<vtkIdList> list = vtkSmartPointer<vtkIdList>::New();
+
+    bool same = false;
+    // check that the point is not already on an existing vertex
+    for (int i = 0; i < 3; i++) {
+      double* p = cell->GetPoints()->GetPoint(i);
+      if (p[0] == pt[0] && p[1] == pt[1] && p[2] == pt[2]) {
+        same = true;
+      }
+    }
+
+    if (!same) {
+      int vertex_id = this->reference_mesh_->GetPoints()->InsertNextPoint(pt);
+
+      list->SetNumberOfIds(3);
+      list->SetId(0, cell->GetPointId(0));
```

Comment:
is the ordering(back face vs front face) correct here? I think it should be:

```
f0 = [0 1 p]
f1 = [1 2 p]
f2 = [2 0 p]
```
(f is face, p is new point, 0 1 2 and old point indices)

Reply:
Yes!  Thank you.  I tossed it in without thinking about it and meant to check it later and forgot.